km/km_signal.c

```diff
@@ -384,7 +393,19 @@ static int km_can_interrupt(km_vcpu_t* vcpu, void* data)
  */
 static void km_interrupt_thread(int signo)
 {
-   km_vcpu_apply_all(km_can_interrupt, (void*)&signo);
+   km_vcpu_t* vcpu;
+
+   if (signo >= 0) {
+      km_signal_lock();
+      if ((vcpu = TAILQ_FIRST(&km_signal_sleep_queue)) != NULL) {
+         TAILQ_REMOVE(&km_signal_sleep_queue, vcpu, signal_link);
```

Comment:
The vcpu is removed from the sleep queue here, then the vcpu blocked in the system call is interrupted.  The system call returns and the vcpu is removed from the sleep queue again.

I don't think TAILQ_REMOVE() will handle this correctly.

Am I misunderstanding this?

Reply:
And I think this needs to be under the if below ...